schannel: fix mTLS with TLS 1.3

kevinAlbs opened the pull request:
# Summary

Import Secure Channel keys as persisted and delete later. Fixes access violation reported in #17626.

# Background & Motivation

Experiments and references suggest persisted keys (rather than ephemeral) are required to support TLS 1.3.

Quoting [.NET runtime](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs#L261-L262):
> on Windows we do not support ephemeral keys.

Quoting [Private Key Lifetime on Windows and the January 2021 .NET Framework Security and Quality Rollup](https://support.microsoft.com/en-us/topic/private-key-lifetime-on-windows-and-the-january-2021-net-framework-security-and-quality-rollup-3f097a10-f798-4ffd-8511-4757ebaf2c70):
> The default behavior is that the private key gets loaded into a persisted (named) key via one of the system cryptography libraries

Quoting https://stackoverflow.com/a/72101855:
> The TLS layer on Windows requires that the private key be written to disk

Quoting [PFXImportCertStore](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-pfximportcertstore):
> if the handshake is performed externally to the calling process in LSASS.exe, it is not possible to use PKCS12_NO_PERSIST_KEY

This PR models the .NET runtime [SafeCertContextHandleWithKeyContainerDeletion](https://github.com/dotnet/runtime/blob/0d628dae1cb5b8d0eb76d9549134edfe30d6219c/src/libraries/Common/src/Microsoft/Win32/SafeHandles/SafeCertContextHandleWithKeyContainerDeletion.cs) to delete the persisted key.

# Testing

This PR adds a regression test. `PFXImportCertStore` may import as a CAPI key or CNG key. On my machine, the test imported as CAPI. To also test CNG key deletion, the test was run by temporarily passing the [flag `PKCS12_PREFER_CNG_KSP`](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-pfximportcertstore) to trigger storing in CNG.

# Caveats

`ncrypt.dll` is newly required for CNG key deletion.

`delete_imported_key` ignores errors.

If the process terminates abnormally, the key may remain on disk.

kevinAlbs committed:
schannel: fix mTLS with TLS 1.3

Fixes #17626

Import key as persisted to fix observe access violation. Delete key when done. Matches .NET behavior.

Add regression test: test4002

Remove from KNOWN_BUGS

kevinAlbs committed:
include `<ncrypt.h>` after `<wincrypt.h>`

To try to fix VS 2010 build

kevinAlbs committed:
add ncrypt.lib refs to old VS solutions

To try to fix VisualStudioSolution build

kevinAlbs committed:
try to fix mingw 4.8.1

I expect `schannel.h` might include `ncrypt.h` indirectly via mingw headers.

kevinAlbs committed:
relocate windows header includes to schannel.h

For consistency with other windows includes

vszakats commented on lib/vtls/schannel.h:
```diff
@@ -36,6 +36,15 @@
+/* Define _WIN32_WINNT to ensure MinGW headers show NCrypt definitions.
+ * MinGW 4.8.1 headers require _WIN32_WINNT>=600 to show  NCrypt symbols.
+ * curl documents Vista as lowest supported Windows version */
+#define _WIN32_WINNT 0x0600  /* Requires Windows Vista */
```
curl currently supports XP as minimum.
Bumping to Vista is scheduled for later via #18009.
A logic like above will need to be done in the build
systems (and config-win32.h), or in the env.

jay commented:
Is there any possibility that this could delete a key that was already in the user's store before libcurl was used?

vszakats commented:
My concern is that the key will be on the disk again, while the process is running.
If the process terminates abnormally, they key may remain on disk. Also, is it
guaranteed that the deletion is done in a secure way?

kevinAlbs committed:
check if ncrypt is available

kevinAlbs committed:
do not require ncrypt.lib for older VS projects

Import as ephemeral instead.

kevinAlbs committed:
import as ephemeral if NCrypt unavailable

This guards against the (unlikely?) scenario:
- curl is built without defining HAVE_NCRYPT. (This may happen if using the older VS 2010, 2012, or 2013 solution files and targeting a Windows supporting CNG keys).
- The target platform is a Windows supporting CNG (Vista+)
- A key is imported into CNG.

Without this guard, the CNG key would be not deleted.

This change continues to import keys as ephemeral if NCrypt is unavailable.

kevinAlbs committed:
Merge branch 'master' into with-failing-test.mtls.schannel

kevinAlbs committed:
use `curlx_malloc` and `curlx_free`

kevinAlbs committed:
Revert "multibyte: limit `curlx_convert_*wchar*()` functions to Unicode builds"

This reverts commit 9517b41b509f1098ed0b3b7ff20b53a566fa3f5e.

vszakats commented on lib/vtls/schannel.c:
```diff
@@ -1476,6 +1489,125 @@ static bool add_cert_to_certinfo(const CERT_CONTEXT *ccert_context,
+      prov_name =
+        curlx_convert_wchar_to_UTF8(prov_info->pwszProvName);
+      #endif
+      CryptAcquireContext(
```
Is CryptAcquireContext() expecting UTF-8 strings?

vszakats commented on lib/curl_config.h.cmake:
```diff
@@ -770,6 +770,9 @@ ${SIZEOF_TIME_T_CODE}
 #cmakedefine USE_SCHANNEL 1
 
+/* to enable Windows CNG APIs available in Vista and above */
+#cmakedefine HAVE_NCRYPT 1
```
The next release will require Vista, and ncrypt can be used unconditionally.
Let's delete these bits.

vszakats commented on lib/vtls/schannel.c:
```diff
@@ -622,12 +631,7 @@ static CURLcode schannel_acquire_credential_handle(struct Curl_cfilter *cf,
-        else
-          cert_store = PFXImportCertStore(&datablob, pszPassword, 0);
-
+        cert_store = PFXImportCertStore(&datablob, pszPassword, import_flags);
```
I take issue with deleting this security feature.
Leaving secret keys on disk and later trying to 
delete them manually seem very error-prone.

kevinAlbs committed:
Merge branch 'master' into with-failing-test.mtls.schannel

kevinAlbs committed:
update KNOWN_BUGS.md

kevinAlbs committed:
include wide char conversion in non-UNICODE build

kevinAlbs committed:
switch `curlx_unicode_free` to `curlx_free`

bagder commented:
This is now > six months old since last update and has many merge conflicts. The provided solution is an unknown and no one seems to report this issue as a problem in the meantime - possibly because mTLS usage with schannel is low.

We cannot accept a change that brings back the old problem with lingering files. Closing this until we can have a fix that does not also revert an even larger problem.
